File: src/AdventOfCode/Puzzles/Y2016/Day01.cs
﻿// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

namespace MartinCostello.AdventOfCode.Puzzles.Y2016;

/// <summary>
/// A class representing the puzzle for <c>https://adventofcode.com/2016/day/1</c>. This class cannot be inherited.
/// </summary>
[Puzzle(2016, 01, "No Time for a Taxicab", RequiresData = true)]
public sealed class Day01 : Puzzle
{
    private static readonly char[] Separators = [',', ' '];

    /// <summary>
    /// An enumeration of directions.
    /// </summary>
    internal enum Direction
    {
        /// <summary>
        /// Turn left.
        /// </summary>
        Left,

        /// <summary>
        /// Turn right.
        /// </summary>
        Right,
    }

    /// <summary>
    /// Gets the number of blocks to the Easter Bunny's headquarters.
    /// </summary>
    public int BlocksToEasterBunnyHQ { get; private set; }

    /// <summary>
    /// Gets the number of blocks to the Easter Bunny's headquarters ignoring duplicates.
    /// </summary>
    public int BlocksToEasterBunnyHQIgnoringDuplicates { get; private set; }

    /// <summary>
    /// Calculate the distance, in blocks, by following the specified instructions.
    /// </summary>
    /// <param name="input">The instructions to follow.</param>
    /// <param name="ignoreDuplicates">Whether to ignore duplicate locations.</param>
    /// <returns>
    /// The shortest distance, in blocks, away from the origin following the instructions would take you.
    /// </returns>
    internal static int CalculateDistance(string input, bool ignoreDuplicates)
    {
        var bearing = CardinalDirection.North;
        var position = Point.Empty;

        var instructions = ParseDirections(input);
        var positions = new List<Point>(instructions.Length);

        foreach (Instruction instruction in instructions)
        {
            bearing = Turn(bearing, instruction.Direction);

            for (int i = 0; i < instruction.Distance; i++)
            {
                if (!ignoreDuplicates)
                {
                    if (positions.Contains(position))
                    {
                        break;
                    }

                    positions.Add(position);
                }

                Size delta = bearing switch
                {
                    CardinalDirection.North => Directions.Down,
                    CardinalDirection.South => Directions.Up,
                    CardinalDirection.East => Directions.Right,
                    CardinalDirection.West => Directions.Left,
                    _ => throw new PuzzleException($"The bearing {bearing} is not known."),
                };

                position += delta;
            }
        }

        return position.ManhattanDistance();
    }

    /// <inheritdoc />
    protected override async Task<PuzzleResult> SolveCoreAsync(string[] args, CancellationToken cancellationToken)
    {
        string instructions = await ReadResourceAsStringAsync(cancellationToken);

        BlocksToEasterBunnyHQIgnoringDuplicates = CalculateDistance(instructions, ignoreDuplicates: true);
        BlocksToEasterBunnyHQ = CalculateDistance(instructions, ignoreDuplicates: false);

        if (Verbose)
        {
            Logger.WriteLine(
                "The Easter Bunny's headquarters are {0:N0} blocks away.",
                BlocksToEasterBunnyHQIgnoringDuplicates);

            Logger.WriteLine(
                "The Easter Bunny's headquarters are {0:N0} blocks away if it is the first location visited twice.",
                BlocksToEasterBunnyHQ);
        }

        return PuzzleResult.Create(BlocksToEasterBunnyHQIgnoringDuplicates, BlocksToEasterBunnyHQ);
    }

    /// <summary>
    /// Parses the directions to the Easter Bunny headquarters from the specified string.
    /// </summary>
    /// <param name="input">The directions to parse.</param>
    /// <returns>
    /// An array of <see cref="Instruction"/> containing the directions to Easter Bunny HQ.
    /// </returns>
    private static Instruction[] ParseDirections(string input)
    {
        string[] instructions = input.Split(Separators, StringSplitOptions.RemoveEmptyEntries);

        var result = new Instruction[instructions.Length];

        for (int i = 0; i < instructions.Length; i++)
        {
            ReadOnlySpan<char> rawInstruction = instructions[i];

            result[i] = new Instruction()
            {
                Direction = rawInstruction[0] == 'L' ? Direction.Left : Direction.Right,
                Distance = Parse<int>(rawInstruction[1..]),
            };
        }

        return result;
    }

    /// <summary>
    /// Returns the new bearing after turning left or right from the specified bearing.
    /// </summary>
    /// <param name="bearing">The current bearing.</param>
    /// <param name="direction">The direction to turn in.</param>
    /// <returns>
    /// The new bearing.
    /// </returns>
    private static CardinalDirection Turn(CardinalDirection bearing, Direction direction)
    {
        return bearing switch
        {
            CardinalDirection.East => direction == Direction.Left ? CardinalDirection.North : CardinalDirection.South,
            CardinalDirection.North => direction == Direction.Left ? CardinalDirection.West : CardinalDirection.East,
            CardinalDirection.South => direction == Direction.Left ? CardinalDirection.East : CardinalDirection.West,
            CardinalDirection.West => direction == Direction.Left ? CardinalDirection.South : CardinalDirection.North,
            _ => throw new PuzzleException($"Invalid bearing '{bearing}'."),
        };
    }

    /// <summary>
    /// A class representing an instruction for movement. This class cannot be inherited.
    /// </summary>
    internal sealed class Instruction
    {
        /// <summary>
        /// Gets or sets the direction to travel.
        /// </summary>
        internal Direction Direction { get; set; }

        /// <summary>
        /// Gets or sets the distance to travel.
        /// </summary>
        internal int Distance { get; set; }
    }
}

File: src/AdventOfCode.Site/scripts/ts/puzzles/2016/Day01.ts
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import { CardinalDirection, Point, Puzzle, Size } from '../index';
import { Puzzle2016 } from './Puzzle2016';

export class Day01 extends Puzzle2016 {
    blocksToEasterBunnyHQ: number;
    blocksToEasterBunnyHQIgnoringDuplicates: number;

    override get name() {
        return 'No Time for a Taxicab';
    }

    override get day() {
        return 1;
    }

    static calculateDistance(input: string, ignoreDuplicates: boolean) {
        let bearing = CardinalDirection.north;
        let position = Point.empty;

        const instructions = this.parseDirections(input);
        const positions: Point[] = [];

        for (const instruction of instructions) {
            bearing = this.turn(bearing, instruction.direction);

            for (let i = 0; i < instruction.distance; i++) {
                if (!ignoreDuplicates) {
                    if (positions.find((other) => other.equals(position))) {
                        break;
                    }

                    positions.push(position);
                }

                let delta: Size;

                switch (bearing) {
                    case CardinalDirection.north:
                        delta = new Size(0, 1);
                        break;
                    case CardinalDirection.south:
                        delta = new Size(0, -1);
                        break;
                    case CardinalDirection.east:
                        delta = new Size(1, 0);
                        break;
                    case CardinalDirection.west:
                        delta = new Size(-1, 0);
                        break;
                    default:
                        throw new Error(`The bearing ${bearing} is not known.`);
                }

                position = Point.add(position, delta);
            }
        }

        return Point.manhattanDistance(position, Point.empty);
    }

    override solveCore(_: string[]) {
        const instructions = this.readResourceAsString();

        this.blocksToEasterBunnyHQIgnoringDuplicates = Day01.calculateDistance(instructions, true);
        this.blocksToEasterBunnyHQ = Day01.calculateDistance(instructions, false);

        console.info(`The Easter Bunny's headquarters are ${this.blocksToEasterBunnyHQIgnoringDuplicates} blocks away.`);
        console.info(
            `The Easter Bunny's headquarters are ${this.blocksToEasterBunnyHQ} blocks away if it is the first location visited twice.`
        );

        return this.createResult([this.blocksToEasterBunnyHQIgnoringDuplicates, this.blocksToEasterBunnyHQ]);
    }

    private static parseDirections(input: string) {
        const instructions = input.split(/[,\s]+/);

        const result: Instruction[] = [];

        for (let i = 0; i < instructions.length; i++) {
            const rawInstruction = instructions[i];

            result[i] = {
                direction: rawInstruction[0] === 'L' ? Direction.left : Direction.right,
                distance: Puzzle.parse(rawInstruction.slice(1)),
            };
        }

        return result;
    }

    private static turn(bearing: CardinalDirection, direction: Direction) {
        switch (bearing) {
            case CardinalDirection.east:
                return direction === Direction.left ? CardinalDirection.north : CardinalDirection.south;
            case CardinalDirection.north:
                return direction === Direction.left ? CardinalDirection.west : CardinalDirection.east;
            case CardinalDirection.south:
                return direction === Direction.left ? CardinalDirection.east : CardinalDirection.west;
            case CardinalDirection.west:
                return direction === Direction.left ? CardinalDirection.south : CardinalDirection.north;
            default:
                throw new Error(`Invalid bearing '${bearing}'.`);
        }
    }
}

enum Direction {
    left,
    right,
}

class Instruction {
    direction: Direction;
    distance: number;
}
